src/libstore/store-api.cc

```diff
@@ -660,9 +688,17 @@ void copyPaths(ref<Store> srcStore, ref<Store> dstStore, const StorePathSet & st
         [&](const Path & storePathS) {
             checkInterrupt();
 
-            auto storePath = dstStore->parseStorePath(storePathS);
+            auto storePath = srcStore->parseStorePath(storePathS);
+            auto info = srcStore->queryPathInfo(storePath);
+
+            auto storePathForDst = storePath.clone();
+            if (info->isContentAddressed(*srcStore)) {
+                storePathForDst = dstStore->makeFixedOutputPathFromCA(storePath.name(), info->ca);
+                if (storePathForDst != storePath)
+                    debug("rewriting path '%s' to '%s' for substituter '%s'", srcStore->printStorePath(storePath), dstStore->printStorePath(storePathForDst), dstStore->getUri());
+            }
```

Comment:
This requires a change to the `copyPaths()` interface, since it can now add a *different* path to the destination store than what the caller asked for, with no way to know what path got added. So maybe it should return a `std::map<StorePath, StorePath>`.